pandas/types/cast.py

```diff
@@ -322,8 +325,52 @@ def infer_dtype_from_scalar(val, pandas_dtype=False):
         whether to infer dtype including pandas extension types.
         If False, scalar belongs to pandas extension types is inferred as
         object
+    downcast : bool, default False
+        If True, downcast float and integer types to the smallest width
+        type that can hold `val`.
+
+    .. versionadded:: 0.20.0
+    allow_uint : bool, default False
+        If True and `downcast` is True, non-negative integers will be
+        downcast to smallest width unsigned integer type that can hold
+        them. Otherwise, signed types are always downcast to signed types
+        and the same for unsigned types.
+
+    .. versionadded:: 0.20.0
     """
 
+    def _downcast_dtype(dtype, val):
+        if 'float' in str(dtype):
+            if ((val > np.finfo(np.float16).min and
+                    val < np.finfo(np.float16).max) or val is np.nan):
+                return np.float16
+            elif (val > np.finfo(np.float32).min and
+                    val < np.finfo(np.float32).max):
```

Comment:
This code depends on `maybe_downcast_to_dtype` which in turn ignores scalar dtypes [here](https://github.com/pandas-dev/pandas/blob/master/pandas/types/cast.py#L53). We have really gotten down in the weeds. Coupling all of these type conversions that all do somewhat similar things is probably generating more debt than just having some repetition in the code base in the first place. 

Thoughts?